AsyncDisplayKit/ASDisplayNode.mm

```diff
@@ -645,28 +625,42 @@ - (BOOL)shouldRasterizeDescendants
 {
   ASDisplayNodeAssertThreadAffinity(self);
   ASDN::MutexLocker l(_propertyLock);
+  ASDisplayNodeAssert(!((_hierarchyState & ASHierarchyStateRasterized) && _flags.shouldRasterizeDescendants),
+                      @"Subnode of a rasterized node should not have redundant shouldRasterizeDescendants enabled");
   return _flags.shouldRasterizeDescendants;
 }
 
-- (void)setShouldRasterizeDescendants:(BOOL)flag
+- (void)setShouldRasterizeDescendants:(BOOL)shouldRasterize
 {
   ASDisplayNodeAssertThreadAffinity(self);
   ASDN::MutexLocker l(_propertyLock);
   
-  if (_flags.shouldRasterizeDescendants == flag)
+  if (_flags.shouldRasterizeDescendants == shouldRasterize)
     return;
   
-  _flags.shouldRasterizeDescendants = flag;
+  _flags.shouldRasterizeDescendants = shouldRasterize;
   
   if (self.isNodeLoaded) {
-    //recursively tear down or build up subnodes
+    // Recursively tear down or build up subnodes.
+    // TODO: When disabling rasterization, preserve rasterized backing store as placeholderImage
+    // while the newly materialized subtree finishes rendering.  Then destroy placeholderImage to save memory.
     [self recursivelyClearContents];
-    [self __tearDown:flag subnodesOfNode:self];
-    if (flag == NO) {
-      [self _addSubnodeViewsAndLayers];
-    }
     
-    [self recursivelyDisplayImmediately];
+    ASDisplayNodePerformBlockOnEverySubnode(self, ^(ASDisplayNode *node) {
+      if (shouldRasterize) {
+        [node enterHierarchyState:ASHierarchyStateRasterized];
+        [node __unloadNode];
+      } else {
+        [node exitHierarchyState:ASHierarchyStateRasterized];
+        [node __loadNode];
```

Comment:
Yes - thanks!  I refactored this code a few times and ended up in a state that is incorrect.  Now fixed, although I don't like how it is factored, but I'm going to revisit it later (since it should work correctly now) to make it shorter / more elegant — running low on time tonight.